src/NUnitFramework/framework/Internal/Reflect.cs

```diff
@@ -67,33 +67,29 @@ public static class Reflect
         /// <returns>The array of methods found</returns>
         public static MethodInfo[] GetMethodsWithAttribute(Type fixtureType, Type attributeType, bool inherit)
         {
-            List<MethodInfo> list = new List<MethodInfo>();
-
             var flags = AllMembers | (inherit ? BindingFlags.FlattenHierarchy : BindingFlags.DeclaredOnly);
-            foreach (MethodInfo method in fixtureType.GetMethods(flags))
-            {
-                if (method.IsDefined(attributeType, inherit))
-                    list.Add(method);
-            }
 
-            list.Sort(new BaseTypesFirstComparer());
-
-            return list.ToArray();
+            return fixtureType.GetMethods(flags)
+                .Where(method => method.IsDefined(attributeType, inherit))
+                .OrderBy(method => method.DeclaringType, BaseTypesFirstComparer.Instance)
+                .ToArray();
         }
 
-        private class BaseTypesFirstComparer : IComparer<MethodInfo>
+        private sealed class BaseTypesFirstComparer : IComparer<Type>
         {
-            public int Compare(MethodInfo m1, MethodInfo m2)
-            {
-                if (m1 == null || m2 == null) return 0;
+            public static BaseTypesFirstComparer Instance { get; } = new BaseTypesFirstComparer();
 
-                Type m1Type = m1.DeclaringType;
-                Type m2Type = m2.DeclaringType;
+            private BaseTypesFirstComparer() { }
 
-                if (m1Type == m2Type) return 0;
-                if (m1Type.IsAssignableFrom(m2Type)) return -1;
+            public int Compare(Type x, Type y)
+            {
+                if (x == y) return 0;
+                if (x is null) return -1;
+                if (y is null) return 1;
```

Comment:
Previously, this method returned `0` when either parameter was null. This would have made the comparison no longer transitive. When A < B, the sorting method could compare and find that A = null and B = null and would conclude that A = B. Inconsistent values returned from this method means undefined behavior for the sorting method.

https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.icomparer-1.compare#remarks

> Comparing `null` with any reference type is allowed and does not generate an exception. A null reference is considered to be less than any reference that is not null.

Reply:
I've just realized that we still have the intransitivity problem. Imagine these types:

```cs
class A { }
class B : A { }
class C { }
```

This comparer should consider A < B, A == C, and B == C. "Base types first" is an intransitive comparison. We can't implement it using any of the BCL sorting methods because they rely on the assumption that `IComparer<>.Compare` is always transitive.

Reply:
Alternatively, we could refrain from getting the methods out of order and having to sort them in the first place. I'll implement this.